RESTClient factory: client-go share custom transports between different group/versions

aojea opened the pull request:
#### What type of PR is this?

/kind bug
/kind cleanup

#### What this PR does / why we need it:

Until now, when creating a Kubernetes Clientset, a new RESTClient was created per group/version. Since RESTClient transport can not be cached in some occasions, like when using a custom transport, this has the consequence that each group/version had an independent TCP connection to the API server.

Since the Transport, and another fields, of the RESTClient can be shared when creating a global Kubernetes Clientset, this PR adds a new RESTClientFactory types that allows to create RESTClients sharing the common fields.

#### Which issue(s) this PR fixes:

Fixes #103224

#### Special notes for your reviewer:

It turns out that some of the imported code in the generated clients is required to register the scheme and can not be generalized, specially the one required for apiextensions that uses its own Codecs

https://github.com/kubernetes/kubernetes/blob/e5976909c6fb129228a67515e0f86336a53884f0/staging/src/k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset/typed/apiextensions/v1beta1/apiextensions_client.go#L23

In order to make the implementation less disruptive, I've added a new constructor (similar to the ones used in the informers) to the generated clientset types, reusing current parametrization and making it totally compatible with current implementation.

#### Does this PR introduce a user-facing change?
```release-note
bugfix: client-go and RESTClient allow to share customized transports for multiple group/versions
```

/sig api-machinery

aojea commented on staging/src/k8s.io/client-go/rest/factory.go:
```diff
@@ -0,0 +1,175 @@
+	}
+}
+
+func (r RESTClientFactory) NewForGV(group, version string) *RESTClient {
```
I couldn't figure out how to pass the gv directly from the generator so I took this brute force approach for the sake of prototyping it

aojea commented on staging/src/k8s.io/code-generator/cmd/client-gen/generators/generator_for_clientset.go:
```diff
@@ -75,7 +75,6 @@ func (g *genClientset) GenerateType(c *generator.Context, t *types.Type, w io.Wr
 	m := map[string]interface{}{
 		"allGroups":                            allGroups,
 		"Config":                               c.Universe.Type(types.Name{Package: "k8s.io/client-go/rest", Name: "Config"}),
-		"DefaultKubernetesUserAgent":           c.Universe.Function(types.Name{Package: "k8s.io/client-go/rest", Name: "DefaultKubernetesUserAgent"}),
```
this doesn't seem to be used anywhere

Commit:
integration test client-go multiple transports

aojea commented:
@liggitt this solves the problem of multiple transports and only changes the constructor for the clientset, avoiding issues with the code that initializes clientsets from the typed objects directly.
However, it fails for the tests with CRDS, I need to figure out why, but PTAL to the approach to see if this makes sense (appreciate if you can give me a hint on the failing tests)

aojea commented on staging/src/k8s.io/client-go/rest/factory.go:
```diff
@@ -0,0 +1,175 @@
+// others: /apis/<group>/<version>
+// "meta.k8s.io" or empty version: /
+func versionedAPIPathFromGV(groupVersion schema.GroupVersion) string {
+	// meta.k8s.io or empty version returns nothing
+	if groupVersion == metav1.SchemeGroupVersion || groupVersion.Version == "" {
+		return ""
+	}
```
I'm not sure if this is correct

aojea commented on staging/src/k8s.io/client-go/rest/factory.go:
```diff
@@ -0,0 +1,175 @@
+	Client *http.Client
+}
+
+func RESTClientFactoryFor(config *Config) (*RESTClientFactory, error) {
```
this is based on the defaulting of current generated types

aojea commented on staging/src/k8s.io/client-go/rest/factory.go:
```diff
@@ -0,0 +1,175 @@
+	}, err
+}
+
+func RESTClientFactoryFromClient(c RESTClient) *RESTClientFactory {
```
this solves this bug https://github.com/kubernetes/kubernetes/issues/103224

aojea commented on test/integration/apiserver/apiserver_test.go:
```diff
@@ -2205,6 +2207,56 @@ func expectPartialObjectMetaV1EventsProtobuf(t *testing.T, r io.Reader, values .
+	}
+	t.Logf("Listed %d deployments on the cluster", len(d.Items))
+
+	if counter != 1 {
```
running this test in current master (without this patch) yields 3 connections here

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset/clientset.go:
```diff
@@ -71,17 +71,16 @@ func NewForConfig(c *rest.Config) (*Clientset, error) {
 	}
 
+	var cs Clientset
+	cs.apiextensionsV1beta1 = apiextensionsv1beta1.New(factory.NewForGV("Apiextensions.k8s.io", "V1beta1"))
```
capital case looks odd

aojea replied:
fixed in https://github.com/kubernetes/kubernetes/pull/105432/commits/ef6516fa9d78b724a27856c75a120e84520dcbe0

aojea commented:
Interesting, restoring the generated code for apiextensions and the non-client-go types work for all the tests
https://github.com/kubernetes/kubernetes/pull/105432/commits/a9f8d5712977e5a6f1581d8470fcc4709e5b94e9

Commit:
RESTClientFactory

RESTClientFactory allow to create RESTClients sharing the same
transport, base url, rateLimiter, warningHandler and createBackoffMgr.

New RESTClients can be created passing the corresponding api path
and the ClientContentConfiguration: GroupVersion, ContentType and
Negotiator used to obtain the encoders and decoders.

A RESTClientFactory can be created from a restClient Config or
directly from an existent RESTClient.

Commit:
Generate Kubernetes Clientset from factory

Currently, the Kubernetes clientset creates a new Client per
type, not allowing to reuse the transport and other common
fields. This is especially problematic when using a custom
Dialer, since the transport can not be cacheable, ending
in as much as connections as RESTClients generated.

Generate the Kubernetes Clientset from a RESTClientFactory, so
all the individual clientset share the same transport.

The code generated follows the same pattern than the previous one,
because each generated typed object imports triggers the machinery
necessary to build and add the schemes.

Commit:
update client code generated with RESTClientFactory

aojea commented on staging/src/k8s.io/code-generator/cmd/client-gen/generators/generator_for_group.go:
```diff
@@ -189,6 +193,18 @@ func NewForConfigOrDie(c *$.restConfig|raw$) *$.GroupGoName$$.Version$Client {
+// NewForFactory creates a new $.GroupGoName$$.Version$Client for the given RESTClientFactory.
+func NewForFactory(f *$.restRESTClientFactory|raw$) *$.GroupGoName$$.Version$Client {
+	var config $.restClientContentConfig|raw$
+	config.GroupVersion = $.SchemeGroupVersion|raw$
+	config.Negotiator = $.runtimeNewClientNegotiator|raw$(scheme.Codecs.WithoutConversion(), $.SchemeGroupVersion|raw$)
```
these 2 imports are the ones that triggers the scheme apimachinery magic

aojea commented:
/assign @sttts @liggitt

aojea commented:
/kind cleanup

liggitt commented on staging/src/k8s.io/client-go/rest/factory.go:
```diff
@@ -0,0 +1,131 @@
+		config.UserAgent = DefaultKubernetesUserAgent()
+	}
+
+	transport, err := TransportFor(config)
+	if err != nil {
+		return nil, err
+	}
```
the transport/client bits are really the only thing we need/want to reuse... the rest could be reconstructed

ideally, we wouldn't duplicate all the other stuff from the RESTClient constructor... it's vulnerable to drift

aojea replied:
transport + baseURL, right?
agree with the rest

aojea replied:
... the problem is that we miss options that are obtained from Config, if you construct the Factory from Config maybe is legit to expect to create the Clients with those

liggitt replied:
baseURL isn't expensive to construct new per RESTClient, the transport is the only thing we care about sharing between all the RESTClients in the clientset, right?

I would try to structure the factory so that it *just* constructs the transport/client, then factor out RESTClient construction so that it does the same, and have a new rest client constructor that takes a config and a client

aojea replied:
maybe I'm overthinking this and we just need something like https://github.com/kubernetes/kubernetes/pull/105490/commits/225e32401e95e73532e2209c3a0eabef0a3482ed ?

liggitt commented on staging/src/k8s.io/client-go/rest/factory.go:
```diff
@@ -0,0 +1,131 @@
+		}
+	}
+
+	// RateLimiter
+	rateLimiter := config.RateLimiter
+	if rateLimiter == nil {
+		qps := config.QPS
+		if config.QPS == 0.0 {
+			qps = DefaultQPS
+		}
+		burst := config.Burst
+		if config.Burst == 0 {
+			burst = DefaultBurst
+		}
+		if qps > 0 {
+			rateLimiter = flowcontrol.NewTokenBucketRateLimiter(qps, burst)
+		}
+	}
```
this will be an observable behavior change, where all subclients of a clientset will share the same rate limiter, right? if we're narrowly trying to address the multiple-transport-construction issue, I don't think we want this

aojea replied:
ahh, is a ratelimiter per group version

liggitt reviewed (commented):
could we scope down the factory to just the client/transport?

Commit:
Add new RESTClient constructor with options

Add a new RESTClient constructor using the functional options pattern
so we can safely add new options in the future without having to
modify the constructor signature or adding new constructors.

Commit:
RESTFactory with options

Commit:
update code generators

Commit:
update generated code

aojea commented on staging/src/k8s.io/client-go/rest/factory.go:
```diff
@@ -97,6 +98,9 @@ func (r RESTClientFactory) NewClientWithOptions(apiPath string, config ClientCon
 		versionedAPIPath: DefaultVersionedAPIPath(apiPath, config.GroupVersion),
 		content:          config,
+		// defaults
+		rateLimiter:      flowcontrol.NewTokenBucketRateLimiter(DefaultQPS, DefaultBurst),
```
it is defaulted on RESTClientFor this way, but I don't have clear if this should be here or in the clientset,

Commit:
remove rate limiter default

aojea commented on staging/src/k8s.io/code-generator/cmd/client-gen/generators/generator_for_clientset.go:
```diff
@@ -144,12 +148,23 @@ func NewForConfig(c *$.Config|raw$) (*Clientset, error) {
 	}
+	options := []$.RESTClientOption|raw${}
+	if configShallowCopy.RateLimiter != nil {
+		options = append(options, $.RESTClientWithRateLimiter|raw$(configShallowCopy.RateLimiter))
```
the problem is that this is sharing the same ratelimiter :/, before each clienset created its own

fedebongio commented:
/cc @yliaog 
/triage accepted
